### ISSUE TITLE
Active Resource Measuring Point

### ISSUE THREAD

**Issue by klinakuf** [PalladioSimulator/Palladio-Analyzer-Slingshot-Extension-PCM-Core#14]
# Schlachtplan 
* introduce new event to publish state of active resource, e.g. `ActiveResourceStateUpdated`:
  - Should extend `AbstractEntityChanged`. the entity is the resource container / queue.
  - Should *not* extend `AbstractGenericEvent` because theres's no need for Reification.
    - thus, should als *not* extend `ModelPassedElement`, event though i actually liked that idea :(
  - Might extend `AbstractJobEvent`, to unify with other job related events
    - con : entity is enforced to `Job`, but for AR changed events, the entity is the resource :/ 
    - alternative : common interface, like with the SEFFProgressed? (i dont want to resort to parametrize all results with `DESEvent` orslt.)
   - Solution: independet, and with 'broader' result, e.g. Result<DESEvent> (like usage)

  * `ActiveResourceStateUpdated` (Design and Hierarchy of this event is imho the only relecant point of discussion.)
    - `entity` : the Resource, maybe interface `ActiveResource`, or even saver `IResource`?
    - `queue length` : `long` 
      - what if multiple CPUs? SimuLizar returns queue per CPU.  
      - does slingshot even consider multiple CPUs? [TODO]
    - `resource demand` : `double` 
      - in seconds, though: what is this even?? 

* @ ResourceSimulation :
  - `@OnEvent(when = JobInitiated.class, then = { JobProgressed.class, ActiveResourceStateUpdated.class }, [...])`
  - `@OnEvent(when = JobFinished.class, then = { ActiveResourceFinished.class, ActiveResourceStateUpdated.class }, [...])`

* @ Active Resources :  
  - always also publish state, if updated.

* @ Monitoring Behaviour
  - `@OnEvent(when = MonitorModelVisited.class, then = CalculatorRegistered.class, cardinality = EventCardinality.SINGLE)`
    - create Probe Sets & initialize Unary (?) Calculators
  - `@OnEvent(when = ActiveResourceStateUpdated.class, then = ProbeTaken.class, cardinality = EventCardinality.SINGLE)`
    - at some point gotta call `Probe::takeMeasurement`

* @ Probes :
  - State of Active Resource : `EventDimensionlessProbe extends EventBasedProbe<Double, Dimensionless>`
    - *hint* : use `Dimensionless.UNIT` as unit
    - `getMeasurement()` returns the queue length extracted from the update event.
  - Resource Demand : `EventOtherTimeProbe extends EventBasedProbe<Double, Duration>`
    - basically the same as `EventCurrentSimulationTimeProbe` 
    - but `getMeasurement()` returns the resource demand (in seconds) from the update event.
  - Probes need a `EventDistinguisher`
    - for passive resource and usage model probes, it's mostly the user's Id. 
    - for AR it should probably be the AR id? 

* @ Calculator : what kind do we need? 
  - Single probe ProbeSet : `DefaultCalculatorProbeSets::createSingularProbeConfiguration(final Probe probe) : CalculatorProbeSet`
  - build Calculator : `IGenericCalculatorFactory::buildCalculator(MetricDescription metric, MeasuringPoint measuringPoint, CalculatorProbeSet probeConfiguration) : Calculator`
    - beware :  for a singular probe set, you get a `UnaryCalculator`. And `UnaryCalculator`s require a `MeasurementListMeasureProvider` as `MeasuringValues`. This is because the `UnaryCalculator` does (usually not) take point in time measurements and thus needs a `MeasuringValues` that is already a tuple of *(point in time, value)*.
    -  for start-stop probe sets, a `BasicMeasurement` (i.e. a *single* value) was sufficient, because the `TimeSpanCalculator` can (and does) use the value of the stop-probe as point in time for the result tuple.
    - TBH, i'm very much not sure how i feel about this design. but i'm pretty sure it's a rather negative emotion :(

**Comment by stiesssh** [comment]
EDIT: for now, this will be state of active resource and resource demand only. as utilisation is moe use full with aggragation, i'd handle utilization in a separate issue.

**Comment by stiesssh** [comment]
### on whether CPU replicas are considered by SimuLizar / Slingshot
* **SimuLizar**
  * ProcSharing : Yes (if looking at resulting responsetimes), this does not match the reported states of active resources though
    * internally, SimuLizar has a distince queue for each replica.
  * FCFS : ignored. result is alway equal to using only 1 replica. 
    * possible replica usage 1 : processing each request with all replicas, thus getting speed up. 
    * possible replica usage 2 : processing each request with one replica, but processing the first n requests in parallel. 
   - they do neither though :( 
* **Slingshot**
  * ProcSharing : Yes. Number of replicas is only a speed up factor though, i.e. the reported states of active resources is always for the *whole resource* not only for the first of the replicated CPUs, as it is with SimuLizar.
  * FCFS : ignored. result is alway equal to using only 1 replica.

**Comment by stiesssh** [comment]
closed by PR PalladioSimulator/Palladio-Analyzer-Slingshot-Extension-PCM-Core#17
